dhcp4relay/src/dhcp4relay_stats.cpp

```diff
@@ -0,0 +1,240 @@
+    {DHCPv4_MESSAGE_TYPE_DISCOVER, "Discover"},
+    {DHCPv4_MESSAGE_TYPE_OFFER, "Offer"},
+    {DHCPv4_MESSAGE_TYPE_REQUEST, "Request"},
+    {DHCPv4_MESSAGE_TYPE_DECLINE, "Decline"},
+    {DHCPv4_MESSAGE_TYPE_ACK, "Acknowledge"},
+    {DHCPv4_MESSAGE_TYPE_NAK, "NegativeAcknowledge"},
+    {DHCPv4_MESSAGE_TYPE_RELEASE, "Release"},
+    {DHCPv4_MESSAGE_TYPE_INFORM, "Inform"},
+    {DHCPv4_MESSAGE_TYPE_MALFORMED, "Malformed"},
+    {DHCPv4_MESSAGE_TYPE_DROP, "Dropped"}};
+
+/**
+ * @code                calculate_delta(uint64_t new_value, uint64_t old_value);
+ *
+ * @brief               Helper function to calculate safe delta with overflow handling.
+ *
+.* @param new_value     new uint64_t value
+.* @param old_value     old uint64_t value
+ *
+ * @return              delta value
+ */
+uint64_t calculate_delta(uint64_t new_value, uint64_t old_value) {
+    if (new_value >= old_value) {
+        return new_value - old_value;
+    } else {
+        // Handle overflow case
+        return (std::numeric_limits<uint64_t>::max() - old_value) + new_value + 1;
+    }
+}
+
+/**
+ * @code                get_counters_data()
+ *
+ * @brief               GET function to fetch data of private member interfaces_cntr_table
+ *
+ * @return              std::unordered_map<std::string, DHCPCounters>
+ */
+std::unordered_map<std::string, DHCPCounters> DHCPCounter_table::get_counters_data() {
+    return interfaces_cntr_table;
+}
+
+/**
+ * @code                DHCPCounter_table::db_update_loop();
+ *
+ * @brief               Loop to update dhcp stats to the DB periodically.
+ *                      This loop is triggered by a new thread which is responsible to update stats to DB.
+ *
+ *                      Mutex lock is taken to copy the data locally, before parsing and setting to DB.
+ *
+ * @return              none
+ */
+void DHCPCounter_table::db_update_loop() {
+    std::shared_ptr<swss::DBConnector> cntrs_db = std::make_shared<swss::DBConnector>("COUNTERS_DB", 0);
+    std::shared_ptr<swss::Table> cntr_table = std::make_shared<swss::Table>(
+        cntrs_db.get(), "COUNTERS_DHCPV4");
+
+    while (!stop_thread) {
+        std::this_thread::sleep_for(std::chrono::seconds(DHCP_RELAY_DB_UPDATE_TIMER_VAL));
+
+        // Copy the data by taking lock for processing and populating to DB
+        std::unordered_map<std::string, DHCPCounters> interfaces_copy;
+        {
+            std::lock_guard<std::mutex> lock(interfaces_mutex);
+            interfaces_copy = interfaces_cntr_table;
+        }
+
+        /* These steps are followed before updating to Redis:
+           1. Fetch present values from redis - existing _fields
+           2. Update counters with 'cache values' + 'existing_feilds'
+           3. Populate to DB
+        */
+        for (const auto& [interface, counters] : interfaces_copy) {
+            // RX counters
+            std::vector<swss::FieldValueTuple> existing_rx_fields;
+            std::string rx_key = interface + "|RX";
```

Comment:
The separator in COUNTERS_DB should be ":" rather than "|" https://github.com/sonic-net/sonic-buildimage/blob/master/dockers/docker-database/database_config.json.j2#L51